lnrpc/invoicesrpc/htlc_modifier.go

```diff
@@ -0,0 +1,110 @@
+package invoicesrpc
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/lightningnetwork/lnd/invoices"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// htlcModifierConfig contains the configuration for an RPC invoice HTLC
+// modifier server.
+type htlcModifierConfig struct {
+	// chainParams is required to properly marshall an invoice for RPC.
+	chainParams *chaincfg.Params
+
+	// serverStream is a bidirectional RPC server stream to send invoices to
+	// a client and receive accept responses from the client.
+	serverStream Invoices_HtlcModifierServer
+
+	// modificationInterceptor is the HTLC modification interceptor that
+	// will be used to intercept and resolve invoice HTLCs.
+	modificationInterceptor invoices.HtlcModifier
+}
+
+// htlcModifier is a helper struct that handles the lifecycle of an RPC invoice
+// HTLC modifier server instance.
+//
+// This struct handles passing send and receive RPC messages between the client
+// and the invoice service.
+type htlcModifier struct {
+	// cfg contains the configuration for the invoice HTLC modifier.
+	cfg htlcModifierConfig
+}
+
+// newHtlcModifier creates a new RPC invoice HTLC modifier handler.
+func newHtlcModifier(cfg htlcModifierConfig) *htlcModifier {
+	return &htlcModifier{
+		cfg: cfg,
+	}
+}
+
+// run sends the intercepted invoice HTLCs to the client and receives the
+// corresponding responses.
+func (r *htlcModifier) run() error {
+	// Register our invoice modifier.
+	r.cfg.modificationInterceptor.SetClientCallback(r.onIntercept)
+	defer r.cfg.modificationInterceptor.SetClientCallback(nil)
+
```

Comment:
following on from my previous comment, this could perhaps instead be:

```
unregister := r.cfg.modificationInterceptor.SetClientCallback(r.onIntercept)
defer unregister()
```

and perhaps `RegisterInterceptor` is maybe a more suitable name than `SetClientCallback`?